### DISCUSSION TITLE
Add php 7.4 Opcache Preloading capability

### DISCUSSION THREAD

**Discussion by Flyingmana** [OpenMage/magento-lts#2151] [Ideas]
## Related Links

* https://wiki.php.net/rfc/preload
* https://dev.to/restoreddev/introduction-to-php-7-4-preload-4lg4
* https://github.com/restoreddev/preloading-example
* https://github.com/symfony/symfony/pull/32032
* https://github.com/symfony/symfony/pull/33689

## Tasks
* [ ] add basic capability to preload a class
* [ ] preload a set of much used Classes
* [ ] introduce a config to allow modules defining own lists of classes for preloading 
* [ ] make sure it does not break with older PHP versions

**Comment by seansan** [comment]
How would we apply this? https://github.com/restoreddev/preloading-example/blob/master/preload.php

Would we preload ALL files? Or start with /lib first? or /app/code?

It seems to me that we could already test this quite easily if someone is running 7.4? Just set the path to the preload.php file and make the file compatible with Magento structure

The questions would then remain
- how do we protect /preload.php or is it placed 1 folder under public_html? (you dont want outsiders to access it)
- would we offer some static variable in preload.php to control which folders we do or dont preload? (see above LIB only, all code or also templates?)
- which folders would we preload, with which include statement *.php + *.phtml? and which ignore?

- also after changing files we would have to restart php, but is there maybe also a refresh pleload routine?

In short: 
- can we already create some preload.php file for M1?
- and where would we safely place it?

**Comment by colinmollenhour** [comment]
The output of Aoe_ClassPathCache would be a good way to create a preload file for testing before doing a full implementation. Compared against a well-tuned opcache (plenty of memory to never have evictions) and running Aoe_ClassPathCache I'm not expecting a huge improvement, though. Hopefully I'm proven wrong. :)

**Comment by Flyingmana** [comment]
We probably dont want to add all files, but often used ones. Probably rather have it configurable as some people prefer memory efficiency, and others full speed, and people need to do benchmarks.


@colinmollenhour the preloading is for avoiding the uncached initial requests, which in some high usage scenarious especially with docker can be a big difference. (I know of one Company, who did still use the M1 Magento Compiler, as it made big differences for some of their customers performance wise as beeing able to cause outages during high traffic scenarious.)

**Comment by seansan** [comment]
From reading the article there is another improvement and that is that the
cache does not have to be validated every request And for every file and
that bit/mem usage improvements can be applied in php internally because
the codes is presumed to be static. Also a cache lookup is not necessary
(is it there) because php assumes it to be there.

Freely interpretated
